engine/src/main/java/org/terasology/rendering/nui/animation/AbstractAnimation.java

```diff
@@ -0,0 +1,151 @@
+    }
+    private AnimState myState;
+
+    public AbstractAnimation() {
+        myListeners = new ArrayList<AnimationListener>();
+        myFrames = new ArrayList<AbstractFrame>();
+        myLastFrame = null;
+        myState = AnimState.STOPPED;
+    }
+
+    public void addFrame(AbstractFrame theFrame) {
+        myFrames.add(theFrame);
+    }
+
+    public void start() {
+        myState = AnimState.RUNNING;
+        for (AnimationListener li : myListeners) {
+            li.onStart();
+        }
+    }
+
+    public void update(float dt) {
+        switch (myState) {
+        case STOPPED:
+            return;
+        case RUNNING: {
+            if (myFrames.size() == 0) {
+                onEnd();
+                return;
+            }
+            myCurrentDuration += dt;
+            if (myCurrentDuration >= myFrames.get(0).getDuration()) {
+                myLastFrame = myFrames.get(0);
+                myFrames.remove(0);
+                return;
+            }
+            break;
+        }
+        }
+    }
+
+    private void onEnd() {
+        myState = AnimState.STOPPED;
+        for (AnimationListener li : myListeners) {
+            li.onEnd(myRepeatCount);
+        }
+    }
+
+    public void addListener(AnimationListener li) {
+        myListeners.add(li);
+    }
+
+    public void removeListener(AnimationListener li) {
+        myListeners.remove(li);
+    }
+
+    public enum RepeatType {
+        REPEAT, REVERSE
+    }
+
+    public interface InterpolatorInterface {
+        /**
+         * Returns where an interpolated value should be based on
+         * where the position an animation is in.
+         *
+         * @param v position of the animation between the start and end [0:1]
+         * Or between [setStart:setEnd] if they have been called.
+         *
+         * @return where the interpolated value should be
+         */
+        float getInterpolation(float v);
+
+        void setStart(float v);
+        float getStart();
+
+        void setEnd(float v);
+        float getEnd();
+    }
+
+    public interface FrameComponentInterface<T> {
+        void setInterpolation(float v, T theFrom, T theTo);
+    }
+
+    public static class AbstractFrame {
+        private List<Object> myFromComponents;
+        private List<Object> myToComponents;
+        private FrameComponentInterface myInterface;
+        private InterpolatorInterface myInterpolation;
+
+        private RepeatType myRepeatType;
+        private int myRepeatCount;
+        private float myStartDelay;
+        private float myDuration;
+        private boolean myStartUseLastFrame;
+
+        public final float getDuration() { return myDuration; }
+    }
+
+    public interface AnimationListener {
+        void onStart();
+        void onFrameStart();
+        void onStep(float v);
+        void onFrameEnd(int repeatCount);
+        void onEnd(int repeatCount);
+    }
+
+    public static class AnimationAdapter implements AnimationListener {
```

Comment:
Ah, that's a good idea then - although you can also use [Java 8 default methods](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html) instead to avoid having to inherit from a non-obvious "default" class.